File: internal/scriptlets/directory/helpers/parseRegexp.ts
const regexpRegexp = /^\/((?:[^\/\\\r\n]|\\.)+)\/([gimsuy]*)$/;

/**
 * Parses a string into a RegExp.
 * @param input The string to parse.
 * @returns The RegExp if the input is a valid RegExp string, null otherwise.
 */
export function parseRegexp(input: string): RegExp | null {
  const match = input.match(regexpRegexp);
  if (match === null) {
    return null;
  }

  return new RegExp(match[1], match[2]);
}

File: internal/networkrules/rulemodifiers/parseregexp.go
package rulemodifiers

import (
	"fmt"
	"regexp"
)

var (
	// regexpRegexp is a regexp that matches a regexp.
	regexpRegexp = regexp.MustCompile(`^/(.+)/i?$`)
)

// parseRegexp parses regulars expressions contained in rule modifiers.
// It returns nil as the first value if the input doesn't look like a regular expression.
func parseRegexp(s string) (*regexp.Regexp, error) {
	match := regexpRegexp.FindStringSubmatch(s)
	if match == nil {
		return nil, nil
	}

	regexpBody := match[1]
	if s[len(s)-1] == 'i' {
		// Filter lists are designed for JS-based applications whose regex flag syntax is incompatible with that of Go.
		// Therefore, we explicitly convert one to another to maintain compatibility.
		regexpBody = "(?i)" + regexpBody
	}

	regexp, err := regexp.Compile(regexpBody)
	if err != nil {
		return nil, fmt.Errorf("compile regexp: %w", err)
	}

	return regexp, nil
}
